README.md
MaxDps Framework


This addon provides support for highlighting spells on action bars and have support for various DPS helper functions. Donations are welcome & appreciated!
 

Modules for legion:

Death Knight - Frost, Unholy
Demon Hunter - Havoc, Vengeance
Druid - Balance, Feral, Guardian
Hunter - Marksmanship, Beast Mastery
Mage - Fire, Frost, Arcane
Monk - Windwalker
Paladin - Retribution
Priest - Shadow
Rogue - Outlaw
Shaman - Elemental, Enhancement
Warlock - Destruction, Demonology, Affliction
Warrior - Fury, Arms, Protection
Recent updates:

7.2.5 - Custom rotations (type /maxdps in chat to see window)
Check out video how it works (Shaman elemental, other classes works in similar way):



If you wish to contribute please contact me. If you wish to become a tester provide logs of your fights.

Action Bars
Action Bars supported:

ElvUI
Bartender4
Blizzard UI
Dominos
Button Forge (update 1.9)
SuperVillain UI (update 1.11)
Please contact me if you wish to have other action bar addon support.


MaxDps/CHANGES.txt
tag 6b27b0b6f5815795d5a0140f2d2f5fdc33db5389 v7.3.0.1
Author:	Pawel <zacdbg@gmail.com>
Date:	Wed Dec 13 00:24:31 2017 +0100

v7.3.0.1 - Small api update

commit 6301cb06bfc960232da619ba2810b1c3eca9d2e2
Author: Pawel <zacdbg@gmail.com>
Date:   Wed Dec 13 00:24:05 2017 +0100

    v7.3.0.1 - Small api update



MaxDps_Rogue/main.lua

-- Outlaw
local _BladeFlurry = 13877;
local _RolltheBones = 193316;
local _TrueBearing = 193359;
local _SharkInfestedWaters = 193357;
local _GhostlyStrike = 196937;
local _CurseoftheDreadblades = 202665;
local _AdrenalineRush = 13750;
local _MarkedforDeath = 137619;
local _RunThrough = 2098;
local _Broadsides = 193356;
local _PistolShot = 185763;
local _Opportunity = 195627;
local _SaberSlash = 193315;
local _Vanish = 1856;
local _Ambush = 8676;
local _CheapShot = 1833;
local _PreyontheWeak = 131511;
local _DeeperStratagem = 193531;
local _JollyRoger = 199603;
local _GrandMelee = 193358;
local _BuriedTreasure = 199600;
local _SliceandDice = 5171;
local _DeathfromAbove = 152150;
local _Vigor = 14983;
local _CombatPotency = 35551;
local _Bloodlust = 2825;
local _Heroism = 32182;
local _TimeWarp = 80353;
local _Ruthlessness = 14161;
local _Sprint = 2983;
local _BetweentheEyes = 199804;
local _Blind = 2094;
local _CloakofShadows = 31224;
local _Riposte = 199754;
local _GrapplingHook = 195457;
local _CannonballBarrage = 185767;
local _KillingSpree = 51690;
local _Feint = 1966;
local _Elusiveness = 79008;
local _CheatDeath = 31230;
local _CrimsonVial = 185311;
local _Stealth = 1784;
local _HiddenBlade = 202753;
local _Garrote = 202753;
local _Kingsbane = 202753;
local _ToxicBlade = 202753;
local _Envenom = 202753;
local _Mutilate = 202753;
local _Rupture = 1943;

-- Subtlety
local _Nightblade = 195452;
local _ShadowBlades = 121471;
local _ShadowDance = 185313;
local _EnvelopingShadows = 206237;
local _Shadowstrike = 185438;
local _Anticipation = 114015;
local _MasterofShadows = 196976;
local _SymbolsofDeath = 212283;
local _GoremawsBite = 209783;
local _Eviscerate = 196819;
local _Backstab = 53;


-- Auras
local _Stealth = 1784;
local _GreenskinsWaterloggedWristcuffs = 209420;
local _MasterAssassinsInitiative = 235022;

-- Talents
local _isGhostlyStrike = false;
local _isMarkedforDeath = false;
local _isAnticipation = false;
MaxDps.Rogue = {};

function MaxDps.Rogue.CheckTalents()
    MaxDps:CheckTalents();
    _isGhostlyStrike = MaxDps:HasTalent(_GhostlyStrike);
    _isMarkedforDeath = MaxDps:HasTalent(_MarkedforDeath);
    _isAnticipation = MaxDps:HasTalent(_Anticipation);
end

function MaxDps:EnableRotationModule(mode)
    mode = mode or 1;
    MaxDps.Description = 'Rogue [Outlaw]';
    MaxDps.ModuleOnEnable = MaxDps.Rogue.CheckTalents;
    if mode == 1 then
        MaxDps.NextSpell = MaxDps.Rogue.Assassination;
    end;
    if mode == 2 then
        MaxDps.NextSpell = MaxDps.Rogue.Outlaw;
    end;
    if mode == 3 then
        MaxDps.NextSpell = MaxDps.Rogue.Subtlety;
    end;
end

function MaxDps.Rogue.Assassination()
    local timeShift, currentSpell, gcd = MaxDps:EndCast();
    local energy = UnitPower('player', SPELL_POWER_ENERGY);
    local combo = GetComboPoints('player', 'target');

    MaxDps:GlowCooldown(_Vendetta, MaxDps:SpellAvailable(_Vendetta, timeShift));
    
    if MaxDps:PersistentAura(_Stealth, timeShift) then
        return _Garrote;
    end
    
    if MaxDps:SpellAvailable(_Rupture, timeShift) and combo == 5 then
        return _Rupture;
    end      
    
    if MaxDps:SpellAvailable(_Garrote, timeShift) then
        return _Garrote;
    end  

    if MaxDps:SpellAvailable(_Kingsbane, timeShift) then
        return _Kingsbane;
    end

    if MaxDps:SpellAvailable(_ToxicBlade, timeShift) then
        return _ToxicBlade;
    end  

    if MaxDps:SpellAvailable(_Envenom, timeShift) and combo >=4 then
        return _Envenom;
    end    

    if MaxDps:SpellAvailable(_Mutilate, timeShift) then
        return _Mutilate;
    end      
    
    if talents[_Nightstalker] and combo == 5 then
        return _Vanish;
    end

end

function MaxDps.Rogue.Outlaw()
    local timeShift, currentSpell, gcd = MaxDps:EndCast();

    local energy = UnitPower('player', SPELL_POWER_ENERGY);
    local combo = GetComboPoints('player', 'target');

    MaxDps:GlowCooldown(_AdrenalineRush, MaxDps:SpellAvailable(_AdrenalineRush, timeShift));
    MaxDps:GlowCooldown(_CurseoftheDreadblades, MaxDps:SpellAvailable(_CurseoftheDreadblades, timeShift));
    MaxDps:GlowCooldown(_KillingSpree, MaxDps:SpellAvailable(_KillingSpree, timeShift));

    if MaxDps:PersistentAura(_Stealth, timeShift) then
        return _Ambush;
    end

    local curse = MaxDps:Aura(_CurseoftheDreadblades, timeShift, 'HARMFUL');

    -- roll the bones auras
    local rb = {
        TB = MaxDps:Aura(_TrueBearing, timeShift + 3),
        SIW = MaxDps:Aura(_SharkInfestedWaters, timeShift + 3),
        JR = MaxDps:Aura(_JollyRoger, timeShift + 3),
        GM = MaxDps:Aura(_GrandMelee, timeShift + 3),
        BS = MaxDps:Aura(_Broadsides, timeShift + 3),
        BT = MaxDps:Aura(_BuriedTreasure, timeShift + 3),
    }
    -- buty, sprint co CD
    local rbCount = 0;
    for k, v in pairs(rb) do
        if v then
            rbCount = rbCount + 1;
        end
    end

    local shouldRoll = not rb.TB and rbCount < 2;

    if shouldRoll and combo >=4 and energy >= 20 then
        return _RolltheBones;
    end

    if _isGhostlyStrike and not MaxDps:TargetAura(_GhostlyStrike, timeShift + 3) and energy > 27 then
        return _GhostlyStrike;
    end

    if _isMarkedforDeath and combo < 2 and MaxDps:SpellAvailable(_MarkedforDeath, timeShift) then
        return _MarkedforDeath;
    end
    
    if combo >= 5 and MaxDps:Aura(_GreenskinsWaterloggedWristcuffs, timeShift) or and MaxDps:Aura(_MasterAssassinsInitiative, timeShift) then
        return _BetweentheEyes;
    end

    if (combo >= 6 or (combo >= 5 and rb.BS)) then
        return _RunThrough;
    end

    if MaxDps:Aura(_Opportunity, timeShift) and combo <= 4 then
        return _PistolShot;
    end

    return _SaberSlash;
end

function MaxDps.Rogue.Subtlety()
    local timeShift, currentSpell, gcd = MaxDps:EndCast();
    local ShadowDancecd, ShadowDancecurrentCharges, ShadowDancemaxCharges = MaxDps:SpellCharges(_ShadowDance, timeShift)
    local energy = UnitPower('player', SPELL_POWER_ENERGY);
    local combo = GetComboPoints('player', 'target');
    
    --Use Shadowstrike as long you are Stealthed and have less than 5 combo points (7 with Anticipation).
    if _isAnticipation and MaxDps:PersistentAura(_Stealth, timeShift) and combo <= 7 then
        return _Shadowstrike;
    elseif MaxDps:PersistentAura(_Stealth, timeShift) and combo <= 5 then
        return _Shadowstrike;
    end

    --1.Maintain Nightblade.
    if MaxDps:SpellAvailable(_Nightblade, timeShift) and not MaxDps:Aura(_Nightblade, timeShift) then
        return _Nightblade;
    end
    
    --2.Activate Shadow Blades if available.
    if MaxDps:SpellAvailable(_ShadowBlades, timeShift) then
        return _ShadowBlades;
    end

    --3.Enter Shadow Dance if you have 2 charges, or there is less than 20 seconds remaining before you get your second charge (3 charges and 30 seconds for Enveloping Shadows).
    --Shadow Dance should be activated as close to Energy cap as possible (80 with Master of Shadows).
    --Use Shadowstrike as long you are Stealthed and have less than 5 combo points (7 with Anticipation).
    if _isEnvelopingShadows and (ShadowDancecurrentCharges = 3 and ShadowDancecd <= 30 and energy >= 80) then
        return _ShadowDance;
    elseif ShadowDancecurrentCharges = 2 and energy >= 80 or (ShadowDancecurrentCharges = 1 and ShadowDancecd <= 20 and energy >= 80) then
        return _ShadowDance;
    end

    --4.Activate the Symbols of Death buff in combination with at least 1 Shadow Dance charge and Death from Above when you are below 60 Energy.
    if MaxDps:SpellAvailable(_SymbolsofDeath, timeShift) and ShadowDancecurrentCharges >= 1 and energy <= 60 then
        return _SymbolsofDeath;
    end
    
    if isDeathfromAbove then
        return _DeathfromAbove;
    end
    
    --5.Activate Vanish and then Shadowstrike when available if you are at or below 3 Combo Points.
    if combo <= 3 then
        return _Vanish;
    end

    --6.Cast Goremaw's Bite Icon Goremaw's Bite if you are not in Stealth or Shadow Dance, at or below 3 Combo Points, and have less than 50 Energy.
    if MaxDps:SpellAvailable(_GoremawsBite, timeShift) and not MaxDps:PersistentAura(_Stealth, timeShift) and not MaxDps:PersistentAura(_ShadowDance, timeShift) and combo <= 3 and energy <= 50 then
        return _GoremawsBite;
    end
        
    --7.Cast Death from Above at 5+ Combo Points. Always combine with Shadow Dance and Symbols of Death.
    if combo >= 5 and MaxDps:Aura(_ShadowDance, timeShift) and MaxDps:Aura(_SymbolsofDeath, timeShift) then
        return _DeathfromAbove;
    end
    
    --8.Cast Eviscerate to spend Combo Points with 5+.
    if combo >= 5
        return _Eviscerate;
    end
    
    --9.Cast Backstab to generate Combo Points (from behind the target whenever possible).
    return _Backstab;

end


MaxDps/core.lua
local AceGUI = LibStub('AceGUI-3.0');
local lsm = LibStub('AceGUISharedMediaWidgets-1.0');
local media = LibStub('LibSharedMedia-3.0');

MaxDps = LibStub('AceAddon-3.0'):NewAddon('MaxDps', 'AceConsole-3.0', 'AceEvent-3.0', 'AceTimer-3.0');

MaxDps.Textures = {
	['Ping'] = 'Interface\\Cooldown\\ping4',
	['Star'] = 'Interface\\Cooldown\\star4',
	['Starburst'] = 'Interface\\Cooldown\\starburst',
};
MaxDps.FinalTexture = nil;

MaxDps.Colors = {
	Info = '|cFF1394CC',
	Error = '|cFFF0563D',
	Success = '|cFFBCCF02',
}

MaxDps.Classes = {
	[1] = 'Warrior',
	[2] = 'Paladin',
	[3] = 'Hunter',
	[4] = 'Rogue',
	[5] = 'Priest',
	[6] = 'DeathKnight',
	[7] = 'Shaman',
	[8] = 'Mage',
	[9] = 'Warlock',
	[10] = 'Monk',
	[11] = 'Druid',
	[12] = 'DemonHunter',
}

local defaultOptions = {
	global = {
		enabled = true,
		disabledInfo = false,
		debugMode = false,
		disableButtonGlow = false,
		onCombatEnter = true,
		texture = 'Interface\\Cooldown\\ping4',
		customTexture = '',
		highlightColor = {
			r = 1,
			g = 1,
			b = 1,
			a = 1
		},
		cooldownColor = {
			r = 0,
			g = 1,
			b = 0,
			a = 1
		},
		interval = 0.15,
		sizeMult = 1.4
	}
}

local options = {
	type = 'group',
	name = 'MaxDps Options',
	inline = false,
	args = {
		general = {
			order = 10,
			name = 'General',
			type = 'group',
			args = {
				enable = {
					order = 10,
					name = 'Enable',
					desc = 'Enables / disables the addon',
					type = 'toggle',
					width = 'full',
					set = function(info, val)
						MaxDps.db.global.enabled = val;
					end,
					get = function(info) return MaxDps.db.global.enabled end
				},
				onCombatEnter = {
					order = 20,
					name = 'Enable upon entering combat',
					desc = 'Automatically enables helper upon entering combat',
					type = 'toggle',
					width = 'full',
					set = function(info, val)
						MaxDps.db.global.onCombatEnter = val;
					end,
					get = function(info) return MaxDps.db.global.onCombatEnter end
				},
				disableButtonGlow = {
					order = 30,
					name = 'Dissable blizzard button glow (experimental)',
					desc = 'Disables original blizzard button glow',
					type = 'toggle',
					width = 'full',
					set = function(info, val)
						MaxDps.db.global.disableButtonGlow = val;
						MaxDps:UpdateButtonGlow();
					end,
					get = function(info) return MaxDps.db.global.disableButtonGlow end
				},
				interval = {
					order = 40,
					name = 'Interval in seconds',
					desc = 'Sets how frequent rotation updates will be. Low value will result in fps drops.',
					type = 'range',
					min = 0.01,
					max = 2,
					set = function(info, val) MaxDps.db.global.interval = val end,
					get = function(info) return MaxDps.db.global.interval end
				},
			}
		},
		debug = {
			order = 30,
			name = 'Debug options',
			type = 'group',
			args = {
				debugMode = {
					order = 10,
					name = 'Enable debug mode',
					desc = 'Enables spammy chat messages (use this when addon does not work for you)',
					type = 'toggle',
					width = 'full',
					set = function(info, val)
						MaxDps.db.global.debugMode = val;
					end,
					get = function(info) return MaxDps.db.global.debugMode end
				},
				disabledInfo = {
					order = 20,
					name = 'Disable info messages',
					desc = 'Enables / disables info messages, if you have issues with addon, make sure to deselect this.',
					type = 'toggle',
					width = 'full',
					set = function(info, val)
						MaxDps.db.global.disabledInfo = val;
					end,
					get = function(info) return MaxDps.db.global.disabledInfo end
				},
			}
		},
		overlay = {
			order = 20,
			name = 'Overlay settings',
			type = 'group',
			args = {
				texture = {
					order = 10,
					type = 'select',
					dialogControl = 'LSM30_Background',
					name = 'Texture',
					width = 'normal',
					desc = 'Sets Highlight texture (changing this requires UI Reload)',
					values = function()
						return MaxDps.Textures;
					end,
					get = function()
						return MaxDps.db.global.texture;
					end,
					set = function(self, val)
						MaxDps.db.global.texture = val;
						MaxDps:ApplyOverlayChanges();
					end,
				},
				customTexture = {
					order = 20,
					name = 'Custom Texture',
					desc = 'Sets Highlight texture, has priority over selected one (changing this requires UI Reload)',
					type = 'input',
					width = 'normal',
					set = function(info, val)
						MaxDps.db.global.customTexture = strtrim(val or '');
						MaxDps:ApplyOverlayChanges();
					end,
					get = function(info) return strtrim(MaxDps.db.global.customTexture or '') end
				},
				highlightColor = {
					order = 30,
					name = 'Highlight color',
					desc = 'Sets Highlight color',
					type = 'color',
					width = 'normal',
					set = function(info, r, g, b, a)
						local c = MaxDps.db.global.highlightColor;
						c.r, c.g, c.b, c.a = r, g, b, a;
						MaxDps:ApplyOverlayChanges();
					end,
					get = function(info)
						local c = MaxDps.db.global.highlightColor;
						return c.r, c.g, c.b, c.a;
					end,
					hasAlpha = true
				},
				cooldownColor = {
					order = 40,
					name = 'Cooldown color',
					desc = 'Sets Cooldown color',
					type = 'color',
					width = 'normal',
					set = function(info, r, g, b, a)
						local c = MaxDps.db.global.cooldownColor;
						c.r, c.g, c.b, c.a = r, g, b, a;
						MaxDps:ApplyOverlayChanges();
					end,
					get = function(info)
						local c = MaxDps.db.global.cooldownColor;
						return c.r, c.g, c.b, c.a;
					end,
					hasAlpha = true
				},
				sizeMult = {
					order = 50,
					name = 'Overlay size multiplier',
					desc = 'Sets how big will be overlay on the button. 1 = exactly the same as button',
					type = 'range',
					width = 'full',
					min = 0.5,
					max = 2,
					set = function(info, val)
						MaxDps.db.global.sizeMult = val;
						MaxDps:ApplyOverlayChanges();
					end,
					get = function(info) return MaxDps.db.global.sizeMult or 1.4 end
				},
			}
		},
		reset = {
			name = 'Reset settings',
			desc = 'Resets settings to default values',
			type = 'execute',
			func = function()
				MaxDps:ResetSettings();
				MaxDps:ApplyOverlayChanges();
			end
		}
	},
}

function MaxDps:ResetSettings()
	self.db:ResetDB();
end

function MaxDps:GetTexture()
	if self.db.global.customTexture ~= '' and self.db.global.customTexture ~= nil then
		self.FinalTexture = self.db.global.customTexture;
		return self.FinalTexture;
	end

	self.FinalTexture = self.Textures[self.db.global.texture];
	if self.FinalTexture == '' or self.FinalTexture == nil then
		self.FinalTexture = 'Interface\\Cooldown\\ping4';
	end

	return self.FinalTexture;
end

function MaxDps:OnInitialize()
	LibStub('AceConfig-3.0'):RegisterOptionsTable('MaxDps', options, { '/maxdpsopts' });
	self.db = LibStub('AceDB-3.0'):New('MaxDpsOptions', defaultOptions);
	self.optionsFrame = LibStub('AceConfigDialog-3.0'):AddToBlizOptions('MaxDps', 'MaxDps');
	self:RegisterChatCommand('maxdps', 'ShowCustomWindow');

	if not self.db.global.customRotations then
		self.db.global.customRotations = {};
	end
end

MaxDps.DefaultPrint = MaxDps.Print;
function MaxDps:Print(...)
	if self.db.global.disabledInfo then
		return;
	end
	MaxDps:DefaultPrint(...);
end

function MaxDps:EnableRotation()
	self:Print(self.Colors.Info .. 'Enabling');

	if self.NextSpell == nil or self.rotationEnabled then
		self:Print(self.Colors.Error .. 'Failed to enable addon!');
		return;
	end
	self:Print(self.Colors.Info .. 'Fetching');
	self.Fetch();

	MaxDps:CheckTalents();
	if self.ModuleOnEnable then
		self.ModuleOnEnable();
	end

	self:EnableRotationTimer();

	self.rotationEnabled = true;
	self:Print(self.Colors.Success .. 'Enabled');
end

function MaxDps:EnableRotationTimer()
	self.RotationTimer = self:ScheduleRepeatingTimer('InvokeNextSpell', self.db.global.interval);
end

function MaxDps:DisableRotation()
	if not self.rotationEnabled then
		return;
	end

	self:DisableRotationTimer();

	self:DestroyAllOverlays();
	self:Print(self.Colors.Info .. 'Disabling');

	self.Spell = nil;
	self.rotationEnabled = false;
end

function MaxDps:DisableRotationTimer()
	if self.RotationTimer then
		self:CancelTimer(self.RotationTimer);
	end
end

function MaxDps:OnEnable()
	self:RegisterEvent('PLAYER_TARGET_CHANGED');
	self:RegisterEvent('PLAYER_TALENT_UPDATE');
	self:RegisterEvent('ACTIONBAR_SLOT_CHANGED');
	self:RegisterEvent('PLAYER_REGEN_DISABLED');
	self:RegisterEvent('PLAYER_ENTERING_WORLD');

	self:RegisterEvent('ACTIONBAR_HIDEGRID');
	self:RegisterEvent('ACTIONBAR_PAGE_CHANGED');
	self:RegisterEvent('LEARNED_SPELL_IN_TAB');
	self:RegisterEvent('CHARACTER_POINTS_CHANGED');
	self:RegisterEvent('ACTIVE_TALENT_GROUP_CHANGED');
	self:RegisterEvent('PLAYER_SPECIALIZATION_CHANGED');
	self:RegisterEvent('UPDATE_MACROS');
	self:RegisterEvent('VEHICLE_UPDATE');

	self:RegisterEvent('UNIT_ENTERED_VEHICLE');
	self:RegisterEvent('UNIT_EXITED_VEHICLE');
	--	self:RegisterEvent('PLAYER_REGEN_ENABLED');

	self:Print(self.Colors.Info .. 'Initialized');
end

function MaxDps:PLAYER_TALENT_UPDATE()
	self:DisableRotation();
end

function MaxDps:UNIT_ENTERED_VEHICLE(event, unit)
	if unit == 'player' and self.rotationEnabled then
		self:DisableRotation();
	end
end

function MaxDps:UNIT_EXITED_VEHICLE(event, unit)
	if unit == 'player' then
		self:InitRotations();
		self:EnableRotation();
	end
end

function MaxDps:PLAYER_ENTERING_WORLD()
	self:UpdateButtonGlow();
end

function MaxDps:PLAYER_TARGET_CHANGED()
	if self.rotationEnabled then
		if (UnitIsFriend('player', 'target')) then
			return;
		else
			self:InvokeNextSpell();
		end
	end
end

function MaxDps:PLAYER_REGEN_DISABLED()
	if self.db.global.onCombatEnter and not self.rotationEnabled then
		self:Print(self.Colors.Success .. 'Auto enable on combat!');
		self:InitRotations();
		self:EnableRotation();
	end
end

function MaxDps:ButtonFetch()
	if self.rotationEnabled then
		if self.fetchTimer then
			self:CancelTimer(self.fetchTimer);
		end
		self.fetchTimer = self:ScheduleTimer('Fetch', 0.5);
	end
end

MaxDps.ACTIONBAR_SLOT_CHANGED = MaxDps.ButtonFetch;
MaxDps.ACTIONBAR_HIDEGRID = MaxDps.ButtonFetch;
MaxDps.ACTIONBAR_PAGE_CHANGED = MaxDps.ButtonFetch;
MaxDps.LEARNED_SPELL_IN_TAB = MaxDps.ButtonFetch;
MaxDps.CHARACTER_POINTS_CHANGED = MaxDps.ButtonFetch;
MaxDps.ACTIVE_TALENT_GROUP_CHANGED = MaxDps.ButtonFetch;
MaxDps.PLAYER_SPECIALIZATION_CHANGED = MaxDps.ButtonFetch;
MaxDps.UPDATE_MACROS = MaxDps.ButtonFetch;
MaxDps.VEHICLE_UPDATE = MaxDps.ButtonFetch;

function MaxDps:InvokeNextSpell()
	-- invoke spell check
	local oldSkill = self.Spell;

	local timeShift, currentSpell, gcd = MaxDps:EndCast();
	self.Spell = self:NextSpell(timeShift, currentSpell, gcd, self.PlayerTalents);

	if (oldSkill ~= self.Spell or oldSkill == nil) and self.Spell ~= nil then
		self:GlowNextSpellId(self.Spell);
		if WeakAuras then
			WeakAuras.ScanEvents('MAXDPS_SPELL_UPDATE', self.Spell);
		end
	end
	if self.Spell == nil and oldSkill ~= nil then
		self:GlowClear();
		if WeakAuras then
			WeakAuras.ScanEvents('MAXDPS_SPELL_UPDATE', nil);
		end
	end
end

function MaxDps:InitRotations()
	self:Print(self.Colors.Info .. 'Initializing rotations');

	local _, _, classId = UnitClass('player');
	local spec = GetSpecialization();
	self.ClassId = classId;
	self.Spec = spec;

	self:LoadCustomRotations();
	if self.CustomRotations[classId] and self.CustomRotations[classId][spec] then
		self.CurrentRotation = self.CustomRotations[classId][spec];
		self.NextSpell = self.CurrentRotation.fn;
		self:Print(self.Colors.Success .. 'Loaded Custom Rotation: ' .. self.CurrentRotation.name);
	else
		self:LoadModule();
	end
end

function MaxDps:LoadModule()
	if self.Classes[self.ClassId] == nil then
		self:Print(self.Colors.Error .. 'Invalid player class, please contact author of addon.');
		return;
	end

	local module = 'MaxDps_' .. self.Classes[self.ClassId];
	local _, _, _, loadable, reason = GetAddOnInfo(module);

	if IsAddOnLoaded(module) then
		self:Print(self.Colors.Info .. self.Description);
		self:EnableRotationModule(self.Spec);
		self:Print(self.Colors.Info .. 'Finished Loading class module');
		return;
	end

	if reason == 'MISSING' or reason == 'DISABLED' then
		self:Print(self.Colors.Error .. 'Could not find class module ' .. module .. ' or it was disabled.');
		return;
	end

	LoadAddOn(module);

	self:EnableRotationModule(self.Spec);
	self:Print(self.Colors.Info .. self.Description);
	self:Print(self.Colors.Info .. 'Finished Loading class module');
end
